src/System.IO.Pipelines/src/System/IO/Pipelines/PipeOperationState.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.IO.Pipelines
+{
+    [DebuggerDisplay("State: {_state}")]
+    internal struct PipeOperationState
+    {
+        private State _state;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void BeginRead()
+        {
+            if ((_state & State.BeginRead) == State.BeginRead)
+            {
+                ThrowHelper.ThrowInvalidOperationException_AlreadyReading();
+            }
+
+            _state &= ~State.EndRead;
+            _state |= State.BeginRead;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void BeginReadTentative()
+        {
+            if ((_state & State.BeginRead) == State.BeginRead)
+            {
+                ThrowHelper.ThrowInvalidOperationException_AlreadyReading();
+            }
+
+            _state &= ~State.EndRead;
+            _state |= State.BeginReadTenative;
```

Comment:
You can just get rid of the BeginReadTenative [sic] state entirely. The only place it's ever referenced is in EndRead() when it's unset. Nothing ever actually checks whether the state is set or not.

You should leave the BeginReadTentative() method in order to throw if BeginRead is set end to unset the EndRead state, but the state itself can be removed.

Reply:
Yeah. You could get rid of the EndRead state or the ReadingTentative state, but not both.